crate_universe/src/metadata.rs

```diff
@@ -319,6 +324,147 @@ impl VendorGenerator {
+            rustc_bin,
+        }
+    }
+
+    /// Computes the set of enabled features for each target triplet for each crate.
+    pub fn generate(
+        &self,
+        manifest_path: &Path,
+        platform_triples: &BTreeSet<String>,
+    ) -> Result<BTreeMap<CrateId, SelectList<String>>> {
+        let manifest_dir = manifest_path.parent().unwrap();
+        let mut crate_features = BTreeMap::<CrateId, BTreeMap<String, BTreeSet<String>>>::new();
+        for target in platform_triples {
+            let output = Command::new(&self.cargo_bin)
+                .current_dir(manifest_dir)
+                .arg("tree")
+                .arg("--locked")
+                .arg("--manifest-path")
+                .arg(manifest_path)
+                .arg("--prefix=none")
+                .arg("--format=|{p}|{f}|")
+                .arg("--color=never")
+                .arg("--target")
+                .arg(target)
+                .env("RUSTC", &self.rustc_bin)
+                .output()
+                .context(format!(
+                    "Error running cargo to compute features for target '{}', manifest path '{}'",
+                    target,
+                    manifest_path.display()
+                ))?;
+            if !output.status.success() {
+                eprintln!("{}", String::from_utf8_lossy(&output.stdout));
+                eprintln!("{}", String::from_utf8_lossy(&output.stderr));
+                bail!(format!("Failed to run cargo tree: {}", output.status))
+            }
+            for (crate_id, features) in
+                parse_features_from_cargo_tree_output(output.stdout.lines())?
+            {
+                crate_features
+                    .entry(crate_id)
+                    .or_default()
+                    .insert(target.to_owned(), features);
+            }
+        }
+        let mut result = BTreeMap::<CrateId, SelectList<String>>::new();
+        for (crate_id, features) in crate_features.into_iter() {
+            let common = features
+                .iter()
+                .fold(
+                    None,
+                    |common: Option<BTreeSet<String>>, (_, features)| match common {
+                        Some(common) => Some(common.intersection(features).cloned().collect()),
+                        None => Some(features.clone()),
+                    },
+                )
+                .unwrap_or_default();
+            let mut select_list = SelectList::default();
+            for (target, fs) in features {
+                if fs != common {
+                    for f in fs {
+                        select_list.insert(f, Some(target.clone()));
+                    }
+                }
+            }
+            for f in common {
+                select_list.insert(f, None);
+            }
+            result.insert(crate_id, select_list);
+        }
+        Ok(result)
+    }
+}
+
+// Parses the output of `cargo tree --format=|{p}|{f}|`. Other flags may be
```

Comment:
nit: Turn this into a docstring. The info feels like it'd be nice to find in docs
```suggestion
/// Parses the output of `cargo tree --format=|{p}|{f}|`. Other flags may be
```

Reply:
Done.